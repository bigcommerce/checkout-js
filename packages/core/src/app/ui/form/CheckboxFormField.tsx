import { FieldProps } from 'formik';
import { kebabCase } from 'lodash';
import React, { FunctionComponent, memo, ReactNode, useCallback } from 'react';

import BasicFormField from './BasicFormField';
import CheckboxInput from './CheckboxInput';
import FormFieldError from './FormFieldError';

export interface CheckboxFormFieldProps {
    additionalClassName?: string;
    disabled?: boolean;
    name: string;
    id?: string;
    labelContent: ReactNode;
    onChange?(isChecked: boolean): void;
}

const CheckboxFormField: FunctionComponent<CheckboxFormFieldProps> = ({
    additionalClassName,
    disabled = false,
    labelContent,
    onChange,
    name,
    id,
}) => {
    const renderField = useCallback(
        ({ field }: FieldProps) => (
            <>
                <CheckboxInput
                    {...field}
                    checked={!!field.value}
                    disabled={disabled}
                    id={id || field.name}
                    label={labelContent}
                />

                <FormFieldError
                    errorId={`${id ?? name}-field-error-message`}
                    name={name}
                    testId={`${kebabCase(name)}-field-error-message`}
                />
            </>
        ),
        [disabled, id, labelContent, name],
    );

    return (
        <BasicFormField
            additionalClassName={additionalClassName}
            name={name}
            onChange={onChange}
            render={renderField}
        />
    );
};

export default memo(CheckboxFormField);
