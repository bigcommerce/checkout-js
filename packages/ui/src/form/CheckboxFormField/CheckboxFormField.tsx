import { FieldProps } from 'formik';
import { kebabCase } from 'lodash';
import React, { FunctionComponent, memo, ReactNode, useCallback } from 'react';

import BasicFormField from '../BasicFormField/BasicFormField';
import CheckboxInput from '../CheckboxInput/CheckboxInput';
import FormFieldError from '../FormFieldError/FormFieldError';

export interface CheckboxFormFieldProps {
    additionalClassName?: string;
    disabled?: boolean;
    name: string;
    id?: string;
    labelContent: ReactNode;
    onChange?(isChecked: boolean): void;
}

const CheckboxFormField: FunctionComponent<CheckboxFormFieldProps> = ({
    additionalClassName,
    disabled = false,
    labelContent,
    onChange,
    name,
    id,
}) => {
    const renderField = useCallback(
        ({ field }: FieldProps) => (
            <>
                <CheckboxInput
                    {...field}
                    checked={!!field.value}
                    disabled={disabled}
                    id={id || field.name}
                    label={labelContent}
                />

                <FormFieldError
                    errorId={`${id ?? name}-field-error-message`}
                    name={name}
                    testId={`${kebabCase(name)}-field-error-message`}
                />
            </>
        ),
        [disabled, id, labelContent, name],
    );

    return (
        <BasicFormField
            additionalClassName={additionalClassName}
            name={name}
            onChange={onChange}
            render={renderField}
        />
    );
};

export default memo(CheckboxFormField);
