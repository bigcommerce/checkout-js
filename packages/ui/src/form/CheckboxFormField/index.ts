export { default as CheckboxFormField } from './CheckboxFormField';
