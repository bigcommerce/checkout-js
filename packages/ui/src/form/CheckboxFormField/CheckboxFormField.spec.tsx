import { mount } from 'enzyme';
import { Formik } from 'formik';
import { noop } from 'lodash';
import React from 'react';

import CheckboxFormField from './CheckboxFormField';

describe('CheckboxFormField', () => {
    it('matches snapshot with rendered output', () => {
        const component = mount(
            <Formik
                initialValues={{ foobar: true }}
                onSubmit={noop}
                render={() => <CheckboxFormField labelContent="Foobar" name="foobar" />}
            />,
        );

        expect(component.html()).toMatchSnapshot();
    });

    it('renders form field with checkbox', () => {
        const component = mount(
            <Formik
                initialValues={{ foobar: true }}
                onSubmit={noop}
                render={() => <CheckboxFormField labelContent="Foobar" name="foobar" />}
            />,
        );

        expect(component.find('input[type="checkbox"]')).toHaveLength(1);

        expect(component.find('label').text()).toBe('Foobar');
    });

    it('sets initial checked value', () => {
        const component = mount(
            <Formik
                initialValues={{ foobar: true }}
                onSubmit={noop}
                render={() => <CheckboxFormField labelContent="Foobar" name="foobar" />}
            />,
        );

        expect(component.find('input[type="checkbox"]').prop('checked')).toBe(true);
    });

    it('updates checked value when clicked', () => {
        const component = mount(
            <Formik
                initialValues={{ foobar: true }}
                onSubmit={noop}
                render={() => <CheckboxFormField labelContent="Foobar" name="foobar" />}
            />,
        );

        component
            .find('input[type="checkbox"]')
            .simulate('change', { target: { checked: false, name: 'foobar' } });

        expect(component.find('input[type="checkbox"]').prop('checked')).toBe(false);
    });
});
