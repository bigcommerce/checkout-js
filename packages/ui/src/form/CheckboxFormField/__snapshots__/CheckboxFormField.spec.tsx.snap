// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CheckboxFormField matches snapshot with rendered output 1`] = `"<div class=\\"form-field\\"><input name=\\"foobar\\" class=\\"form-checkbox optimizedCheckout-form-checkbox\\" id=\\"foobar\\" type=\\"checkbox\\" value=\\"true\\" checked=\\"\\"><label for=\\"foobar\\" class=\\"form-label optimizedCheckout-form-label\\">Foobar</label></div>"`;
